### ISSUE TITLE
Improve Third-Party Deployment Plugins Reliability and Compatibility

### ISSUE THREAD

**Issue by DimedS** [kedro-org/kedro#4318]
## Overview
Part of #4317. This issue focuses on addressing user challenges with Kedro's third-party connection plugins for deployment. We aim to ensure that these plugins remain reliable and compatible with the latest Kedro versions while offering a flexible approach to managing platform-specific features.

## User Insights and Challenges
- **Kubeflow Plugin**: Users are exploring ways to use the Kubeflow plugin for compiling and deploying Kedro pipelines but face issues with outdated functionality.
- **Databricks Integration**: Some users prefer interacting with the Databricks API manually via Asset Bundles to achieve greater flexibility, rather than using the Kedro-Databricks plugin.
- **Argo Workflows**: Translating Kedro pipelines into Argo Workflows manifests is a significant challenge.
- **SageMaker Plugin**: The Kedro-SageMaker plugin was not used due to incompatibility with Kedro 0.19, limiting its application for certain projects.

Each connection plugin includes two key functionalities:
- Converting Kedro pipelines for different platforms
- Handling API interactions to upload and execute the converted pipelines

## Goals
**Ensure Long-Term Compatibility**: Maintain and update third-party plugins regularly to align with the latest Kedro releases.

## Possible Solutions
- **Split Plugin Functionality**: Divide each plugin's functionality into two parts:
  - **Pipeline Conversion**: Kedro maintainers will support and maintain the part responsible for converting Kedro pipelines to various platforms.
  - **API Interactions**: The community will handle the API interaction component, responsible for uploading and running the converted pipelines on specific platforms.
- **Resource Optimization**: By splitting responsibilities, Kedro maintainers can focus on the core functionality (pipeline conversion), saving resources while ensuring essential compatibility, and empowering the community to maintain API-specific integrations.

# Summary of Tech Design - Dec 11, 2024

The priority for Kedro maintainers should focus on improving the deployment experience for three main cloud platforms: **Sagemaker**, **AzureML**, and **Google VertexAI**.  
- **Short-term goal**: Ensure these plugins are compatible with the latest version of Kedro through collaboration with GetInData and participation of Kedro maintainers in development.  
- **Long-term priorities**:
  - Explore the overall deployment experience on cloud platforms to better guide users and inform future development of Kedro and its plugins.
  - Implement changes in Kedro to simplify and strengthen the development of deployment plugins.
  - Investigate @deepyaman's idea to create a universal deployment product (asset) that Kedro can leverage, potentially eliminating the need for individual deployment plugins.

# Next Steps
- [x] [Update the `kedro-sagemaker` plugin to ensure compatibility with the latest version of Kedro](https://github.com/kedro-org/kedro/issues/4463)

**Comment by DimedS** [comment]
For those who don’t have access to Notion:
<img width="1162" alt="Screenshot 2024-12-09 at 12 56 39" src="https://github.com/user-attachments/assets/120736dc-b2db-440f-94c7-41f67e68ac56">

**Comment by DimedS** [comment]
**Discussion with Get In Data (GiD) - December 9**  
On December 9, we discussed the maintenance of deployment plugins with Get In Data (GiD), the maintainers of six Kedro deployment plugins. Notes from the discussion were documented by Juan Luis and can be found [here](https://www.notion.so/mckinsey/Kedro-GetInData-1573043f5dcf80c194ddc9d15d2f5ed5).

- GiD is willing to provide limited support only for the VertexAI and AzureML plugins due to business priorities, resource constraints, and the dependency on their internal infrastructure for end-to-end testing (they cannot fully open development for these plugins to the community).
- GiD recommends deprecating the `kedro-airflow-k8s` plugin, as there are now better approaches for achieving similar functionality.
- For the remaining plugins - Sagemaker, Snowflake and Kubeflow - GiD is open to granting access to a few Kedro maintainers to assist with updates and PR reviews.
- GiD is currently seeking a new team member and plans to allocate one day per week of their time exclusively to Kedro plugin-related tasks.
- According to GiD, maintaining plugins to stay consistent with Kedro is not particularly challenging. However, ensuring compatibility with other libraries, such as MLflow, can be more complex.

**As the next step, we will determine the level of involvement of Kedro maintainers in the process of maintaining deployment plugins.**

**Comment by DimedS** [comment]
# Summary of Tech Design - Dec 11, 2024

### Discussed
- Reviewed the current state of deployment plugins ([Notion Link](https://www.notion.so/mckinsey/1423043f5dcf80b7a6cdf21901ff55c2?v=7958cdee2674405b82de8dd318d9fa1f)).
- Summarized the outcomes of the conversation with GetInData on December 9.

### Decided
The priority for Kedro maintainers should focus on improving the deployment experience for three main cloud platforms: **Sagemaker**, **AzureML**, and **Google VertexAI**.  
- **Short-term goal**: Ensure these plugins are compatible with the latest version of Kedro through collaboration with GetInData and participation of Kedro maintainers in development.  
- **Long-term priorities**:
  - Explore the overall deployment experience on cloud platforms to better guide users and inform future development of Kedro and its plugins.
  - Implement changes in Kedro to simplify and strengthen the development of deployment plugins.
  - Investigate @deepyaman's idea to create a universal deployment product (asset) that Kedro can leverage, potentially eliminating the need for individual deployment plugins.

### Next Steps
- Coordinate with GetInData to provide Kedro maintainers access to plugin repositories.
- Update the `kedro-sagemaker` plugin to ensure compatibility with the latest version of Kedro.

Feel free to share your thoughts here and suggest any corrections if there's anything inaccurate in the summary.

**Comment by astrojuanlu** [comment]
Added https://github.com/kedro-org/kedro/issues/4463 as sub-issue of this one.

The only other P0 (urgent) action item is to make `kedro-azureml` compatible with latest Kedro version as well. We'll decide about this in the coming weeks.

Remaining action items related to Kedro <> Databricks are mostly about documentation (see https://github.com/kedro-org/kedro/issues/3360#issuecomment-2501355256)

For now, other concrete action items ("kedro deployer", better integration with Argo) are of lesser priority. We will revisit this in a few months.

cc @merelcht

**Comment by astrojuanlu** [comment]
kedro-azureml 0.9.0 works with Kedro 0.19 🎉 https://github.com/getindata/kedro-azureml/releases/tag/0.9.0

**Comment by merelcht** [comment]
Is there any more work to be done here or can we close this? From the way I see it we've accomplished "Improve Third-Party Deployment Plugins Reliability and Compatibility".

**Comment by DimedS** [comment]
Yes, we successfully updated the deployment plugins and documentation, so everything is now in good shape.

<img width="576" alt="Image" src="https://github.com/user-attachments/assets/6dd33975-ff36-4b7e-96c6-b0a1217d38ef" />

https://www.notion.so/mckinsey/1423043f5dcf80b7a6cdf21901ff55c2?v=7958cdee2674405b82de8dd318d9fa1f

The only remaining task is to incorporate the functionality of the deprecated `kedro-airflow-k8s` plugin into `kedro-airflow` ([issue #1025](https://github.com/kedro-org/kedro-plugins/issues/1025)). However, we agreed to defer this until after the 1.0 release, so I believe we can go ahead and close that issue for now.

**Comment by astrojuanlu** [comment]
Agreed, let's close for now!
